### PR TITLE
[STORM-3691] Refactor Resource Aware Strategies.

### PR DESCRIPTION
**Pull request by bipinprasad**
## What is the purpose of the change

*Refactor BaseResourceAwareStrategy, DefaultResourceAwareStrategy and GenericResourceAwareStrategy to remove duplicated code and incompatibilities.*

## How was the change tested

*New test classes to check backward compatibility and modifications to current tests.*

### CODE REVIEW

File: storm-client/src/jvm/org/apache/storm/Config.java [hunk 1]
```diff
@@ -355,12 +362,30 @@
...
     public static final String TOPOLOGY_RAS_CONSTRAINT_MAX_STATE_SEARCH = "topology.ras.constraint.max.state.search";
     /**
+     * The maximum number of states that will be searched looking for a solution in resource aware strategies, e.g.
+     * in BaseResourceAwareStrategy. Backward compatibility config value for old topologies.
+     * @deprecated please use {@link Config#TOPOLOGY_RAS_CONSTRAINT_MAX_STATE_SEARCH}
+     */
+    @IsInteger
+    @IsPositiveNumber
+    @Deprecated
+    public static final String TOPOLOGY_RAS_CONSTRAINT_MAX_STATE_TRAVERSAL = "topology.ras.constraint.max.state.traversal";
```
Comment (agresch) [inline storm-client/src/jvm/org/apache/storm/Config.java:379]: same comment on these....

File: storm-client/src/jvm/org/apache/storm/Config.java [hunk 2]
```diff
@@ -346,6 +345,14 @@
      */
     @IsExactlyOneOf(valueValidatorClasses = { ListOfListOfStringValidator.class, RasConstraintsTypeValidator.class })
     public static final String TOPOLOGY_RAS_CONSTRAINTS = "topology.ras.constraints";
+    /**
+     * Declare scheduling constraints for a topology.
+     * @deprecated please use TOPOLOGY_RAS_CONSTRAINTS.
+     */
+    @Deprecated
+    @CustomValidator(validatorClass = ListOfListOfStringValidator.class)
+    public static final String TOPOLOGY_CONSTRAINTS = "topology.constraints";
```
[the remark this answers is not in the document]
Comment (bipinprasad) [reply storm-client/src/jvm/org/apache/storm/Config.java:354]: Will remove this. No point adding a new deprecated variable.

File: storm-client/src/jvm/org/apache/storm/Config.java (hunk 1 above)
Comment (bipinprasad) [reply storm-client/src/jvm/org/apache/storm/Config.java:379]: Will remove this also. No point adding a new deprecated variable.

File: storm-client/src/jvm/org/apache/storm/Config.java [hunk 3]
```diff
@@ -355,12 +355,21 @@
...
     @IsPositiveNumber
     public static final String TOPOLOGY_RAS_CONSTRAINT_MAX_STATE_SEARCH = "topology.ras.constraint.max.state.search";
     /**
+     * Declare max traversal depth for find solutions that satisfy constraints.
+     * @deprecated please use {@link Config#TOPOLOGY_RAS_CONSTRAINT_MAX_STATE_SEARCH}
+     */
+    @IsInteger
+    @IsPositiveNumber
+    @Deprecated
+    public static final String TOPOLOGY_CONSTRAINTS_MAX_DEPTH_TRAVERSAL = "topology.constraints.max.depth.traversal";
```
Comment (agresch) [inline storm-client/src/jvm/org/apache/storm/Config.java:371]: is this another new deprecation?

Comment (bipinprasad) [reply storm-client/src/jvm/org/apache/storm/Config.java:371]: Removing.

File: storm-server/src/main/java/org/apache/storm/scheduler/resource/strategies/scheduling/IStrategy.java [hunk 4]
```diff
@@ -19,21 +19,29 @@
 
 /**
  * An interface to for implementing different scheduling strategies for the resource aware scheduling.
+ * Scheduler should call {@link #prepare(Map)} followed by {@link #schedule(Cluster, TopologyDetails)}.
+ * <p>
+ *     A fully functioning implementation is in the abstract class {@link BaseResourceAwareStrategy}.
+ *     Subclasses classes should extend {@link BaseResourceAwareStrategy#BaseResourceAwareStrategy(boolean, ObjectResourceSortType)}
```
[the remark this answers is not in the document]
Comment (bipinprasad) [reply storm-server/src/main/java/org/apache/storm/scheduler/resource/strategies/scheduling/IStrategy.java:25]: Fixed documentation.

File: storm-server/src/main/java/org/apache/storm/scheduler/resource/strategies/scheduling/BaseResourceAwareStrategy.java [hunk 5]
```diff
@@ -50,709 +37,308 @@
...
-        return spouts;
+        return true;
     }
 
+    /**
+     * If this config is set to true, unassigned executors will be sorted by topological order with network proximity needs.
+     * @deprecated Use {@link Config#TOPOLOGY_RAS_ORDER_EXECUTORS_BY_PROXIMITY_NEEDS} instead.
+     */
+    @Deprecated
+    public static final String EXPERIMENTAL_TOPOLOGY_RAS_ORDER_EXECUTORS_BY_PROXIMITY_NEEDS
```
[the remark this answers is not in the document]
Comment (bipinprasad) [reply storm-server/src/main/java/org/apache/storm/scheduler/resource/strategies/scheduling/BaseResourceAwareStrategy.java:339]: Removed

**Review by Ethanlm — APPROVED** [review]

**Review by kishorvpatil — APPROVED** [review]
👍

**Pull request merged**